src/Control/Carrier/State/Church.hs

```diff
@@ -23,64 +23,75 @@ module Control.Carrier.State.Church
 ) where
 
 import Control.Algebra
-import Control.Applicative (Alternative(..))
+import Control.Applicative (Alternative(..), liftA2)
 import Control.Effect.State
-import Control.Monad (MonadPlus(..))
+import Control.Monad (MonadPlus)
 import Control.Monad.Fail as Fail
 import Control.Monad.Fix
 import Control.Monad.IO.Class
 import Control.Monad.Trans.Class
 
--- | Run a 'State' effect starting from the passed value.
+-- | Run a 'State' effect starting from the passed value, applying a continuation to the final state and result.
 --
 -- @
--- 'runState' s ('pure' a) = 'pure' (s, a)
+-- 'runState' k s ('pure' a) = k s a
 -- @
 -- @
--- 'runState' s 'get' = 'pure' (s, s)
+-- 'runState' k s 'get' = k s s
 -- @
 -- @
--- 'runState' s ('put' t) = 'pure' (t, ())
+-- 'runState' k s ('put' t) = k t ()
 -- @
 --
 -- @since 1.1.0.0
-runState :: forall s m a . Applicative m => s -> StateC s m a -> m (s, a)
-runState s (StateC m) = m (\ a s -> pure (s, a)) s
+runState :: forall s m a b . (s -> a -> m b) -> s -> StateC s m a -> m b
+runState f s (StateC m) = m f s
```

Comment:
It’s not super important to me that different carriers for the same effect provide exactly the same interface w.r.t. their handlers, and this definition of `runState` for the Church-encoded `StateC` is much more flexible.